proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8855,52 +8778,170 @@ lemma active_sc_tcb_at_budget_sufficient:
-  apply (drule_tac x=target in bspec)
-   apply (clarsimp dest!: set_takeWhileD)
-  apply (intro conjI)
-      apply (clarsimp simp: tcb_at_kh_simps)
-     apply simp
-    apply (clarsimp simp: not_in_release_q_def in_queue_2_def)
-    apply (fastforce dest: dropWhile_takeWhile_distinct)
-   apply (clarsimp dest!: set_takeWhileD)
-   apply (subst (asm) fun_of_m_get_tcb_refill_budget_ready; clarsimp simp: tcb_at_kh_simps vs_all_heap_simps)
-  apply (clarsimp elim!: active_sc_tcb_at_budget_sufficient)
+  (is "valid ?pre _ _")
+  apply (clarsimp simp: awaken_def)
+  apply (rule_tac I="\<lambda>_ s. ?pre s" in valid_whileLoop; (solves simp)?)
+  apply (clarsimp simp: valid_sched_def)
+  apply (rule_tac Q="\<lambda>s. valid_ready_qs s \<and> valid_release_q s \<and> ready_or_release s
+                         \<and> ct_not_in_q s \<and> valid_sched_action s \<and> ct_in_cur_domain s
+                         \<and> valid_blocked s \<and> valid_idle_etcb s \<and> released_ipc_queues s
+                         \<and> active_reply_scs s \<and> active_sc_valid_refills s \<and> valid_idle s
+                         \<and> release_queue s \<noteq> [] \<and> budget_ready (hd (release_queue s)) s"
+               in hoare_weaken_pre[rotated])
+   apply clarsimp
+   apply (prop_tac "read_release_q_non_empty_and_ready s = Some True")
+    using valid_release_q_read_release_q_non_empty_and_ready_bound apply force
+   apply (clarsimp dest!: read_tcb_refill_ready_SomeD read_sc_refill_ready_SomeD
+                          read_tcb_obj_ref_SomeD read_sched_context_SomeD
+                          read_release_q_non_empty_and_ready_SomeD
+                    simp: vs_all_heap_simps)
+  apply (intro hoare_vcg_conj_lift_pre_fix
+         ; (solves \<open>wpsimp simp: awaken_body_def tcb_release_dequeue_def\<close>)?)
+     apply (wpsimp wp: awaken_body_valid_ready_qs)
+    apply (wpsimp wp: possible_switch_to_not_it_ct_not_in_q simp: awaken_body_def tcb_release_dequeue_def)
+    apply (clarsimp simp: valid_release_q_def valid_idle_def vs_all_heap_simps pred_tcb_at_def
+                          obj_at_def)
+    apply (fastforce dest: hd_in_set)
+   apply (wpsimp wp: awaken_body_valid_sched_action)
+  apply (wpsimp wp: possible_switch_to_valid_blocked tcb_release_remove_valid_blocked_except
+              simp: awaken_body_def tcb_release_dequeue_def)
   done
 
 end
 
 crunches awaken
-for cur_tcb[wp]: cur_tcb
-and budget_ready: "\<lambda>s. P (budget_ready t s)"
-and budget_sufficient: "\<lambda>s. P (budget_sufficient t s)"
-and budget_ready_ct: "\<lambda>s. P (budget_ready (cur_thread s) s)"
-and budget_sufficient_ct: "\<lambda>s. P (budget_sufficient (cur_thread s) s)"
-and active_sc_tcb_at'[wp]: "\<lambda>s. P (active_sc_tcb_at t s)"
-and active_sc_tcb_at_ct[wp]: "\<lambda>s. P (active_sc_tcb_at(cur_thread s) s)"
-  (wp: hoare_drop_imps mapM_x_wp')
+  for cur_tcb[wp]: cur_tcb
+  and budget_ready: "\<lambda>s. P (budget_ready t s)"
```

Comment:
Is it intentional that the `budget_` invariants are not `[wp]`?

Reply:
These weren't `[wp]` before. I'll check to see whether they are wp rules, and if so, I'll add them